src/core/Akka.Streams/Dsl/Restart.cs

```diff
@@ -0,0 +1,490 @@
+                Backoff();
+            }
+
+            protected override void StartGraph()
+            {
+                var sourceOut = CreateSubOutlet(_stage.In);
+                Source.FromGraph(sourceOut.Source).RunWith(_stage.SinkFactory(), SubFusingMaterializer);
+            }
+
+            protected override void Backoff()
+            {
+                SetHandler(_stage.In, () =>
+                {
+                    // do nothing
+                });
+            }
+        }
+    }
+
+    /// <summary>
+    /// A RestartFlow wraps a <see cref="Flow"/> that gets restarted when it completes or fails.
+    /// They are useful for graphs that need to run for longer than the <see cref="Flow"/> can necessarily guarantee it will, for
+    /// example, for <see cref="Flow"/> streams that depend on a remote server that may crash or become partitioned. The
+    /// RestartFlow ensures that the graph can continue running while the <see cref="Flow"/> restarts.
+    /// </summary>
+    public static class RestartFlow
+    {
+        /// <summary>
+        /// Wrap the given <see cref="Flow"/> with a <see cref="Flow"/> that will restart it when it fails or complete using an exponential
+        /// backoff.
+        /// This <see cref="Flow"/> will not cancel, complete or emit a failure, until the opposite end of it has been cancelled or
+        /// completed.Any termination by the <see cref="Flow"/> before that time will be handled by restarting it. Any termination
+        /// signals sent to this <see cref="Flow"/> however will terminate the wrapped <see cref="Flow"/>, if it's running, and then the <see cref="Flow"/>
+        /// will be allowed to terminate without being restarted.
+        /// The restart process is inherently lossy, since there is no coordination between cancelling and the sending of
+        /// messages. A termination signal from either end of the wrapped <see cref="Flow"/> will cause the other end to be terminated,
+        /// and any in transit messages will be lost. During backoff, this <see cref="Flow"/> will backpressure.
+        /// This uses the same exponential backoff algorithm as <see cref="Akka.Pattern.Backoff"/>.
+        /// </summary>
+        /// <param name="flowFactory">A factory for producing the <see cref="Flow"/>] to wrap.</param>
+        /// <param name="minBackoff">Minimum (initial) duration until the child actor will started again, if it is terminated</param>
+        /// <param name="maxBackoff">The exponential back-off is capped to this duration</param>
+        /// <param name="randomFactor">After calculation of the exponential back-off an additional random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay. In order to skip this additional delay pass in `0`.</param>
+        public static Flow<TIn, TOut, NotUsed> WithBackoff<TIn, TOut>(Func<Flow<TIn, TOut, NotUsed>> flowFactory, TimeSpan minBackoff, TimeSpan maxBackoff, double randomFactor)
```

Comment:
`flowFactory` should have a signature `Creator[Flow[In, Out, _]`, I don't know how to do it in C#